### PR TITLE
Add AsJob Implementation to Common Code and apply to Compute cmdlets

### PR DESCRIPTION
**Pull request by markcowl**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by shawns1** [comment]
@azuresdkci test this please

File: src/ResourceManager/Profile/Profile.sln [hunk 1]
```diff
@@ -34,6 +34,12 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Common.Network", "
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Common.Storage", "..\..\Common\Commands.Common.Storage\Commands.Common.Storage.csproj", "{65C3A86A-716D-4E7D-AB67-1DB00B3BF72D}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Resources", "..\Resources\Commands.Resources\Commands.Resources.csproj", "{E1F5201D-6067-430E-B303-4E367652991B}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Resources.Rest", "..\Resources\Commands.ResourceManager\Cmdlets\Commands.Resources.Rest.csproj", "{8058D403-06E3-4BED-8924-D166CE303961}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Insights", "..\Insights\Commands.Insights\Commands.Insights.csproj", "{DEA446A1-84E2-46CC-B780-EB4AFDE2460E}"
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Profile.sln]: @markcowl have we checked if the `AzureRM.Insights` module can be removed from `ProfileController`?

File: src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            cmdlet.Fail = true;
+            var job = cmdlet.ExecuteAsJob("Test Job") as AzureLongRunningJob<AzureStreamTestCmdlet>;
+            int times = 0;
+            while (times++ < 20 && job.StatusMessage != "Completed" && job.StatusMessage != "Failed")
+            {
+                Thread.Sleep(TimeSpan.FromSeconds(1));
+            }
+
+            Assert.Equal("Failed", job.StatusMessage);
+            Assert.Equal(2, job.Error.Count);
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs]: @markcowl what do the streams look like for a non-completed job? Are they something that we should validate?

File: src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            Assert.Collection(job.Output, t => Assert.Equal(Output, t.BaseObject));
+            Assert.Collection(job.Error, t => Assert.Equal(Error, t));
+        }
+
+        public class TestCmdletOutput
+        {
+            public string Property { get { return "PropertyValue"; } }
+        }
+
+        public class AzureStreamTestCmdlet : AzurePSCmdlet
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs]: @markcowl is there a way for us to mock out a cmdlet that implements `AzureRmLongRunningCmdlet` so we can write tests to validate that providing the `AsJob` parameter will run the cmdlet as a job, and not providing the parameter will run the cmdlet normally?

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Action/SetAzureVMCommand.cs [hunk 4]
```diff
@@ -70,34 +70,29 @@ public class SetAzureVMCommand : VirtualMachineBaseCmdlet
 
         [Parameter(
             Mandatory = true,
-            Position = 2,
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Action/SetAzureVMCommand.cs]: @markcowl we should let Compute know about these changes (both to `SwitchParameter` properties and the addition of `AsJob`) so they can update their generator accordingly

**Comment by cormacpayne** [comment]
@markcowl also, we should update the help files for each of the cmdlets where `AsJob` was added

File: src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs (hunk 3 above)
Comment (markcowl) [reply src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs]: Yes, but there isn't a good way to run ProcessRecord in a mocked test.  I think the best way to cover this is to cover the extension method that dewtermines whether to run this way, and then a scenario test.

Meanwhile, the eitire set of scenario tests validates that cmdlets with AsJob operate normally.

File: src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs (hunk 2 above)
Comment (markcowl) [reply src/ResourceManager/Profile/Commands.Profile.Test/LongRunningCmdletTests.cs]: We do have tests of stopped job, but it is difficult for a mocked test to reliably check for the state of the fields in between.

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Action/SetAzureVMCommand.cs (hunk 4 above)
Comment (markcowl) [reply src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Action/SetAzureVMCommand.cs]: We have let them know, but yes, we should let them know explicitly how to use this - meanwhile they have told us to edit the generated code in these cases.

File: src/Common/Commands.Common/AzureLongRunningJob.cs [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+            }
+
+            return returnValue as U;
+        }
+
+        /// <summary>
+        /// Run a cmdlet execution as a Job - this is intended to run in a background thread
+        /// </summary>
+        /// <param name="state">The Job record that will track the progress of this job </param>
+        public virtual void RunJob(object state)
```
[the remark this answers is not in the document]
Comment (markcowl) [reply src/Common/Commands.Common/AzureLongRunningJob.cs]: No, it is a required part opf the delegate

**Comment by markcowl** [comment]
@cormacpayne Added issue for test automation here: https://github.com/Azure/azure-powershell/issues/5077

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/235

**Pull request merged**